GNX/man/man8/shutdown.8
.ig
	@(#)shutdown.8	1.2	8/11/83
	@(#)Copyright (C) 1983 by National Semiconductor Corp.
..
.TH SHUTDOWN 8
.UC 4
.SH NAME
shutdown \- close down the system at a given time
.SH SYNOPSIS
.B /etc/shutdown
[
.B \-k
] [
.B \-r
] [
.B \-h
]
time [ warning-message ... ]
.SH DESCRIPTION
.I Shutdown
provides an automated shutdown procedure which a super-user
can use to notify users
nicely when the system is shutting down, saving them from
system administrators, hackers, and gurus, who would otherwise
not bother with niceties.
.LP
.I Time
is the time at which
.I shutdown
will bring the system down and
may take two formats:
.BR + number
and
.RB hour : min.
The first form brings the system down in
.I number
minutes
and the second brings the system down at the time of day indicated
(as a 24\-hour clock).
.PP
At intervals which get closer together as apocalypse approaches,
warning messages are displayed at the terminals of all users on the
system.  Five minutes before shutdown, or immediately if
shutdown is in less than 5 minutes, logins are disabled by
creating
/etc/nologin
and writing a message there.
If this file exists when a user logs in,
.IR login (1)
prints its contents
and exits.
The file is removed just before
.I shutdown
exits.
.PP
At shutdown time a
message is written in the file /usr/adm/shutdownlog, containing the
time of shutdown, who ran shutdown and the reason.
Then a terminate signal is sent at
.I init
to bring the system down to single-user state.
Alternatively, if
.B -r,
.B -h,
or
.B -k
was used, then
.I shutdown
will exec
.IR reboot (8),
.IR halt (8),
or avoid shutting the system down (respectively).
(If it isn't obvious,
.B \-k
is to make people
.I think
the system is going down!)
.PP
The time of the shutdown and the warning message
are placed in /etc/nologin and should be used to
inform the users about when the system will be back up
and why it is going down (or anything else).
.SH FILES
.DT
/etc/nologin	tells login not to let anyone log in
.br
/usr/adm/shutdownlog	log file for succesful shutdowns.
.SH "SEE ALSO"
login(1), reboot(8)
.SH BUGS
Only allows you to kill the system between now and 23:59 if
you use the absolute time for shutdown.
.PP
Times to shutdown are not nice and round, i.e. ``shutdown in 18 seconds''.


GNX/src/cmd/shutdown.c

/*
 * shutdown.c: version 1.1 of 3/15/83
 * 
 */
# ifdef SCCS
static char *sccsid = "@(#)shutdown.c	1.1 (NSC) 3/15/83";
# endif

#include <stdio.h>
#include <ctype.h>
#include <signal.h>
#include <utmp.h>
#include <time.h>
#include <sys/types.h>
/*
 *	/etc/shutdown when [messages]
 *
 *	allow super users to tell users and remind users
 *	of iminent shutdown of unix
 *	and shut it down automatically
 *	and even reboot or halt the machine if they desire
 *
 *		Ian Johnstone, Sydney, 1977
 *		Robert Elz, Melbourne, 1978
 *		Peter Lamb, Melbourne, 1980
 *		William Joy, Berkeley, 1981
 *		Michael Toy, Berkeley, 1981
 *		Dave Presotto, Berkeley, 1981
 */
#ifdef DEBUG
#define LOGFILE "shutdown.log"
#else
#define LOGFILE "/usr/adm/shutdownlog"
#endif
#define	REBOOT	"/etc/reboot"
#define	HALT	"/etc/halt"
#define MAXINTS 20
#define	HOURS	*3600
#define MINUTES	*60
#define SECONDS
#define NLOG		20		/* no of args possible for message */
#define	NOLOGTIME	5 MINUTES
#define IGNOREUSER	"sleeper"

int	do_nothing();
time_t	getsdt();

extern	char *ctime();
extern	struct tm *localtime();

struct	utmp utmp;
int	sint;
int	stogo;
char	tpath[] =	"/dev/";
int	nlflag = 1;		/* nolog yet to be done */
int	killflg = 1;
int	reboot = 0;
int	halt = 0;
char	term[sizeof tpath + sizeof utmp.ut_line];
char	tbuf[BUFSIZ];
char	nolog1[] = "\n\nNO LOGINS: System going down at %5.5s\n\n";
char	*nolog2[NLOG+1];
#ifdef	DEBUG
char	nologin[] = "nologin";
#else
char	nologin[] = "/etc/nologin";
#endif
int slots;
struct interval {
	int stogo;
	int sint;
} interval[] = {
	4 HOURS,	1 HOURS,
	2 HOURS,	30 MINUTES,
	1 HOURS,	15 MINUTES,
	30 MINUTES,	10 MINUTES,
	15 MINUTES,	5 MINUTES,
	10 MINUTES,	5 MINUTES,
	5 MINUTES,	3 MINUTES,
	2 MINUTES,	30 SECONDS,
	40 SECONDS,	10 SECONDS,
	0 SECONDS,	0 SECONDS
};
char *shutter, *getlogin();
main(argc,argv)
	int argc;
	char **argv;
{
	register i, ufd;
	register char **mess, *f;
	char *ts;
	long sdt;
	int h, m;
	long nowtime;
	FILE *termf;

	shutter = getlogin();
	argc--, argv++;
	while (argc > 0 && (f = argv[0], *f++ == '-')) {
		while (i = *f++) switch (i) {
		case 'k':
			killflg = 0;
			continue;
		case 'r':
			reboot = 1;
			continue;
		case 'h':
			halt = 1;
			continue;
		default:
			fprintf(stderr, "shutdown: '%c' - unknown flag\n", i);
			exit(1);
		}
		argc--, argv++;
	}
	if (argc < 1) {
		printf("Usage: %s [ -krh ] shutdowntime [ message ]\n",
		    argv[0]);
		finish();
	}
	if (geteuid()) {
		fprintf(stderr, "NOT super-user\n");
		finish();
	}
	sdt = getsdt(argv[0]);
	argc--, argv++;
	i = 0;
	while (argc-- > 0)
		if (i < NLOG)
			nolog2[i++] = *argv++;
	nolog2[i] = NULL;
	nowtime = time((long *)0);
	m = ((stogo = sdt - nowtime) + 30)/60;
	h = m/60; 
	m %= 60;
	ts = ctime(&sdt);
	printf("Shutdown at %5.5s (in ", ts+11);
	if (h > 0)
		printf("%d hour%s ", h, h != 1 ? "s" : "");
	printf("%d minute%s) ", m, m != 1 ? "s" : "");
#ifndef DEBUG
	signal(SIGHUP, SIG_IGN);
	signal(SIGQUIT, SIG_IGN);
	signal(SIGINT, SIG_IGN);
#endif
	signal(SIGTERM, finish);
	signal(SIGALRM, do_nothing);
	nice(-20);
	fflush(stdout);
#ifndef DEBUG
	if (i = fork()) {
		printf("[pid %d]\n", i);
		exit(0);
	}
#endif
	sint = 1 HOURS;
	f = "";
	for (;;) {
		for (i = 0; stogo <= interval[i].stogo && interval[i].sint; i++)
			sint = interval[i].sint;
		if (stogo <= NOLOGTIME && nlflag) {
			nlflag = 0;
			nolog(sdt);
		}
		if (sint >= stogo || sint == 0)
			f = "FINAL ";
		ufd = open("/etc/utmp",0);
		nowtime = time((long *) 0);
		while (read(ufd,&utmp,sizeof utmp)==sizeof utmp)
		if (utmp.ut_name[0] &&
		    strncmp(utmp.ut_name, IGNOREUSER, sizeof(utmp.ut_name))) {
			strcpy(term, tpath);
			strncat(term, utmp.ut_line, sizeof utmp.ut_line);
			alarm(3);
#ifdef DEBUG
			if ((termf = stdout) != NULL)
#else
			if ((termf = fopen(term, "w")) != NULL)
#endif
			{
				alarm(0);
				setbuf(termf, tbuf);
				fprintf(termf, "\n\n");
				warn(termf, sdt, nowtime);
				if (sdt - nowtime > 1 MINUTES)
					for (mess = nolog2; *mess; mess++)
						fprintf(termf, "%s ", *mess);
				fputc('\n', termf);
				alarm(5);
#ifdef DEBUG
				fflush(termf);
#else
				fclose(termf);
#endif
				alarm(0);
			}
		}
		if (stogo < 0) {
	printf("\n\007\007System shutdown time has arrived\007\007\n");
			log_entry(sdt);
			unlink(nologin);
			if (!killflg) {
				printf("but you'll have to do it yourself\n");
				finish();
			}
#ifndef DEBUG
			if (reboot)
				execle(REBOOT, "reboot", 0, 0);
			if (halt)
				execle(HALT, "halt", 0, 0);
			kill(1, SIGTERM);	/* sync */
			kill(1, SIGTERM);	/* sync */
			sleep(20);
#else
			printf("EXTERMINATE EXTERMINATE\n");
#endif
			finish();
		}
		stogo = sdt - time((long *) 0);
		if (stogo > 0)
			sleep(sint<stogo ? sint : stogo);
		stogo -= sint;
	}
}

time_t
getsdt(s)
register char *s;
{
	time_t t, t1, tim;
	register char c;
	struct tm *lt;

	if (*s == '+') {
		++s; 
		t = 0;
		for (;;) {
			c = *s++;
			if (!isdigit(c))
				break;
			t = t * 10 + c - '0';
		}
		if (t <= 0)
			t = 5;
		t *= 60;
		tim = time((long *) 0) + t;
		return(tim);
	}
	t = 0;
	while (strlen(s) > 2 && isdigit(*s))
		t = t * 10 + *s++ - '0';
	if (*s == ':')
		s++;
	if (t > 23)
		goto badform;
	tim = t*60;
	t = 0;
	while (isdigit(*s))
		t = t * 10 + *s++ - '0';
	if (t > 59)
		goto badform;
	tim += t; 
	tim *= 60;
	t1 = time((long *) 0);
	lt = localtime(&t1);
	t = lt->tm_sec + lt->tm_min*60 + lt->tm_hour*3600;
	if (tim < t || tim >= (24*3600)) {
		/* before now or after midnight */
		printf("That must be tomorrow\nCan't you wait till then?\n");
		finish();
	}
	return (t1 + tim -t);
badform:
	printf("Bad time format\n");
	finish();
}

warn(term, sdt, nowtime)
	FILE *term;
	long sdt, nowtime;
{
	char *ts;
	register delay = sdt - nowtime;

	if (delay > 8)
		while (delay % 5)
			delay++;

	if (shutter)
		fprintf(term,
		    "\007\007*** System shutdown message from %s ***\r\n",
		    shutter);
	else
		fprintf(term,
		    "\007\007*** System shutdown message ***\r\n");
	ts = ctime(&sdt);
	if (delay> 10 MINUTES)
		fprintf(term, "System going down at %5.5s\r\n", ts+11);
	else if ( delay > 60 SECONDS ) {
		fprintf(term, "System going down in %d minute%s\r\n",
		(delay+30)/60, (delay+30)/60 != 1 ? "s" : "");
	} else if ( delay > 0 ) {
		fprintf(term, "System going down in %d second%s\r\n",
		delay, delay != 1 ? "s" : "");
	} else
		fprintf(term, "System going down IMMEDIATELY\r\n");
}

nolog(sdt)
	long sdt;
{
	FILE *nologf;
	register char **mess;

	if ((nologf = fopen(nologin, "w")) != NULL) {
		fprintf(nologf, nolog1, (ctime(&sdt)) + 11);
		putc('\t', nologf);
		for (mess = nolog2; *mess; mess++)
			fprintf(nologf, " %s", *mess);
		putc('\n', nologf);
		fclose(nologf);
	}
}

finish()
{
	signal(SIGTERM, SIG_IGN);
	unlink(nologin);
	exit(0);
}

do_nothing()
{

	signal(SIGALRM, do_nothing);
}

/*
 * make an entry in the shutdown log
 */

char *days[] = {
	"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"
};

char *months[] = {
	"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep",
	"Oct", "Nov", "Dec"
};

log_entry(now)
time_t now;
{
	register FILE *fp;
	register char **mess;
	struct tm *tm, *localtime();

	tm = localtime(&now);
	fp = fopen(LOGFILE, "a");
	if (fp==0)
		return;
	fseek(fp, 0L, 2);
	fprintf(fp, "%02d:%02d  %s %s %2d, %4d.  Shutdown:", tm->tm_hour,
		tm->tm_min, days[tm->tm_wday], months[tm->tm_mon],
		tm->tm_mday, tm->tm_year + 1900);
	for (mess = nolog2; *mess; mess++)
		fprintf(fp, " %s", *mess);
	if (shutter)
		fprintf(fp, " (by %s)", shutter);
	fputc('\n', fp);
	fclose(fp);
}
